libamqpprox/amqpprox_maybesecuresocketadaptor.h

```diff
@@ -331,6 +371,50 @@ class MaybeSecureSocketAdaptor {
                                                              handler);
         }
 
+        if (!d_alarmed && d_dataRateAlarm.remainingQuota() == 0) {
+            if (d_dataRateTimerStarted) {
+                LOG_INFO << "Data Rate Alarm: Hit "
+                         << d_dataRateAlarm.getQuota() << " bytes/s";
+
+                // TODO: Better limit alarm debouncing?
+                d_alarmed = true;
+            }
+            else {
+                // We have hit our quota but we haven't started the refresh
+                // timer yet so it's probable the usage has never reset. Start
+                // the refresh timer for this connection and continue until we
+                // hit it a second time.
+
+                this->onTimer(boost::system::error_code());
+            }
+        }
+
+        if (d_dataRateLimit.remainingQuota() == 0) {
+            if (d_dataRateTimerStarted) {
+                d_dataRateTimer.cancel();
+                d_dataRateTimer.expires_at(d_dataRateTimer.expiry());
```

Comment:
This seems out of line with the description of the PR as I understood it. It seems like this is resetting the exceeded limit on the next 1 second clock tick, which could be anything in the range 0.0s-1.0s, rather than waiting for a full 1.0 second peroid after hitting a limit.

Reply:
Awkward. I believe I started out with waiting a whole second each time but changed to this due to failing to hit expected performance test results (not that unsurprising). I just collected some data on throughput results when being limited, and it looks like the method in code is a lot more predictable - which I think is preferable here. I'll update the PR description to reflect the current behaviour - but equally let me know if you prefer waiting the whole second.

Reply:
No, I was not happy with the 1 second approach so I went looking at the code. I think this approach is better but I am still somewhat concerned about the jitter of confirm times it might introduce, and I wonder if there's a way to smooth the rate limiting out without being too expensive to implement (it doesn't look like there's any leaky bucket or similar implemantation in boost).

If we do decide to take this approach, is the cost of the timers expensive? Because if they might be, then you could have a single 1-second clock and just maintain a map of the sockets that are being limited and need calling back to do the socket read on the next cycle.

Reply:
Yeah as you say - the timers aren't free - the cost of ASIO figuring out which [timers are ready](https://github.com/chriskohlhoff/asio/blob/fc901d79c11c72a0ba81764820d3c6a788c0f075/asio/include/asio/detail/impl/epoll_reactor.ipp#L583) does start to show in `perf` results, but the CPU usage benchmarks I did down below seem to indicate this isn't overly expensive. Do you agree? We could switch the implementation to have a single timer which runs every second - I think that would work, it just needs a bit more thought around socket ownership & how many timers are needed to keep latency / cpu spikes to a minimum. At least at the moment a lower hanging performance fruit is tidying up statistics collection when there are 20k connections (there's some perf output kicking around internally for this).

Reply:
and I think this is everything you already know, but to make it explicit: if someone is pushing a lot of data to the broker they will see increased confirm times when they get throttled, but worth noting that don't do anything to hold back broker confirms flowing to them. 

We could introduce a smoother threshold to help bursty clients which would like to push (for example) large amounts of data in only one second every 30.

I could go for supporting something like `X limit every second, but allow up to Y to accumulate if unused` - which looks roughly like what [`tc` exposes in one of its modes](https://wiki.archlinux.org/title/advanced_traffic_control#Token_Bucket_Filter_(TBF)). We did discuss something a bit more complicated than this for connection rate limits, but eventually decided to go for just a simple limit to keep it easy for users to understand when they get rate limited.

I'm an unsure how much of a difference it would make overall though since 1 second seems like only a medium period of time for a (mostly) asynchronous application like rabbitmq. At the moment I don't think we have a great way to measure if this would make a difference - not to say that means we should stick with what is currently in the PR - just makes it a bit trickier to decide.